aten/src/ATen/native/transformers/cuda/attention.cu

```diff
@@ -952,33 +953,46 @@ std::tuple<at::Tensor, at::Tensor> _efficient_attention_forward(
       output_accum = at::empty(
           {B, M, num_heads, Kv},
           query.options().dtype(
-              TypeTraits<typename Kernel::output_accum_t>::atScalarType()));
+              CutlassToAtenDtype<
+                  typename Kernel::output_accum_t>::atScalarType()));
       p.output_accum_ptr =
           (typename Kernel::output_accum_t*)output_accum.data_ptr();
     } else {
       p.output_accum_ptr = nullptr;
     }
     p.output_ptr = (typename Kernel::output_t*)res.data_ptr();
 
-    if (cu_seqlens_q.has_value()) {
-      p.cu_seqlens_q_ptr = (int32_t*)cu_seqlens_q->data_ptr();
-      p.cu_seqlens_k_ptr = (int32_t*)cu_seqlens_k->data_ptr();
+    if (seqstart_q.has_value()) {
+      p.seqstart_q_ptr = (int32_t*)seqstart_q->data_ptr();
+      p.seqstart_k_ptr = (int32_t*)seqstart_k->data_ptr();
     }
 
-#define ASSIGN_CHECK_OVERFLOW(A, B)                                            \
-  {                                                                            \
-    A = B;                                                                     \
-    TORCH_CHECK(B < std::numeric_limits<decltype(A)>::max(), #B " overflows"); \
-  }
-
-    p.scale = sdp::calculate_scale(query, scale).as_float_unchecked();
     p.num_heads = num_heads;
     p.head_dim = query.size(3);
     p.head_dim_value = value.size(3);
     p.num_queries = max_seqlen_q;
     p.num_keys = max_seqlen_k;
-    p.num_batches = cu_seqlens_q.has_value() ? cu_seqlens_q->size(0) - 1 : B;
-    p.causal = causal;
+    p.num_batches = seqstart_q.has_value() ? seqstart_q->size(0) - 1 : B;
+    p.custom_mask_type = custom_mask_type;
+    p.causal_diagonal_ptr = nullptr;
+    if (causal_diagonal.has_value()) {
+      CHECK_NOSPARSE_LASTCONTIGUOUS_CUDA(causal_diagonal.value());
+      TORCH_CHECK(causal_diagonal->scalar_type() == at::ScalarType::Int);
+      p.causal_diagonal_ptr = (int32_t*)causal_diagonal->data_ptr();
+    }
+
+    p.seqlen_k_ptr = nullptr;
+    if (seqlen_k.has_value()) {
+      CHECK_NOSPARSE_LASTCONTIGUOUS_CUDA(seqlen_k.value());
+      TORCH_CHECK(seqlen_k->scalar_type() == at::ScalarType::Int);
+      p.seqlen_k_ptr = (int32_t*)seqlen_k->data_ptr();
+    }
+
+    if (scale.has_value()) {
+      p.scale = float(*scale);
+    } else {
+      p.scale = float(1.0 / std::sqrt(float(p.head_dim)));
```

Comment:
I still don't like this and wish it was done at the highest level so that every kernel across every device will forever be aligned on what is the default. This is a bad thing to mess up, especially if maybe later on we add more broadcasting support or non-contiguous input support.